python/lsst/daf/butler/registry/obscore/_manager.py

```diff
@@ -0,0 +1,358 @@
+
+    @classmethod
+    def initialize(
+        cls,
+        db: Database,
+        context: StaticTablesContext,
+        *,
+        universe: DimensionUniverse,
+        config: Mapping,
+        datasets: Type[DatasetRecordStorageManager],
+        dimensions: DimensionRecordStorageManager,
+    ) -> ObsCoreTableManager:
+        # Docstring inherited from base class.
+        config_data = Config(config)
+        obscore_config = ObsCoreManagerConfig.parse_obj(config_data)
+
+        schema = ObsCoreSchema(config=obscore_config, datasets=datasets)
+        table = context.addTable(obscore_config.table_name, schema.table_spec)
+        if obscore_config.collection_type is ConfigCollectionType.TAGGED:
+            # Configuration validation guarantees that there is exactly one
+            # collection for TAGGED type.
+            assert obscore_config.collections is not None, "Collections must be defined"
+            return _TaggedObsCoreTableManager(
+                db=db,
+                table=table,
+                schema=schema,
+                universe=universe,
+                config=obscore_config,
+                tagged_collection=obscore_config.collections[0],
+                dimensions=dimensions,
+            )
+        elif obscore_config.collection_type is ConfigCollectionType.RUN:
+            return _RunObsCoreTableManager(
+                db=db,
+                table=table,
+                schema=schema,
+                universe=universe,
+                config=obscore_config,
+                dimensions=dimensions,
+            )
+        else:
+            raise ValueError(f"Unexpected value of collection_type: {obscore_config.collection_type}")
+
+    def config_json(self) -> str:
+        """Dump configuration in JSON format.
+
+        Returns
+        -------
+        json : `str`
+            Configuration serialized in JSON format.
+        """
+        return json.dumps(self.config.dict())
+
+    @classmethod
+    def currentVersion(cls) -> Optional[VersionTuple]:
+        # Docstring inherited from base class.
+        return _VERSION
+
+    def schemaDigest(self) -> Optional[str]:
+        # Docstring inherited from base class.
+        return None
+
+
+class _TaggedObsCoreTableManager(ObsCoreLiveTableManager):
+    """Implementation of ObsCoreTableManager which is used for
+    ``collection_type=TAGGED``.
+    """
+
+    def __init__(
+        self,
+        *,
+        db: Database,
+        table: sqlalchemy.schema.Table,
+        schema: ObsCoreSchema,
+        universe: DimensionUniverse,
+        config: ObsCoreManagerConfig,
+        tagged_collection: str,
+        dimensions: DimensionRecordStorageManager,
+    ):
+        super().__init__(
+            db=db,
+            table=table,
+            schema=schema,
+            universe=universe,
+            config=config,
+            dimensions=dimensions,
+        )
+        self.tagged_collection = tagged_collection
```

Comment:
I started with that idea, but then decided to split it into two classes to avoid that mixed/exclusive state and bunch of `if`s and annoying `mypy` asserts. I can put it back together, should not be too hard.

Reply:
If you've already tried it, please stick with this, because you already know better than I do how the alternative would look.